File: FBSimulatorControl/Configuration/FBSimulatorConfiguration+CoreSimulator.m
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

#import "FBSimulatorConfiguration+CoreSimulator.h"

#import <FBControlCore/FBControlCore.h>

#import <CoreSimulator/SimDevice.h>
#import <CoreSimulator/SimDeviceType.h>
#import <CoreSimulator/SimRuntime.h>

#import <objc/runtime.h>

#import "FBSimulatorError.h"
#import "FBSimulatorServiceContext.h"

@implementation FBSimulatorConfiguration (CoreSimulator)

#pragma mark Matching Configuration against Available Versions

+ (FBOSVersion *)newestAvailableOSForDevice:(FBDeviceType *)device
{
  return [[[[FBSimulatorConfiguration supportedOSVersionsForDevice:device] reverseObjectEnumerator] allObjects] firstObject];
}

- (instancetype)newestAvailableOS
{
  FBOSVersion *os = [FBSimulatorConfiguration newestAvailableOSForDevice:self.device];
  NSAssert(os, @"Expected to be able to find any runtime for device %@", self.device);
  return [self withOSNamed:os.name];
}

+ (FBOSVersion *)oldestAvailableOSForDevice:(FBDeviceType *)device
{
  return [[FBSimulatorConfiguration supportedOSVersionsForDevice:device] firstObject];
}

- (instancetype)oldestAvailableOS
{
  FBOSVersion *os = [FBSimulatorConfiguration oldestAvailableOSForDevice:self.device];
  NSAssert(os, @"Expected to be able to find any runtime for device %@", self.device);
  return [self withOSNamed:os.name];
}

+ (instancetype)inferSimulatorConfigurationFromDevice:(SimDevice *)simDevice error:(NSError **)error
{
  FBOSVersionName osName = simDevice.runtime.name;
  FBOSVersion *osVersion = FBiOSTargetConfiguration.nameToOSVersion[osName];
  if (!osVersion) {
    return [[FBSimulatorError
      describeFormat:@"Could not obtain OS Version for %@, perhaps it is unsupported by FBSimulatorControl", osName]
      fail:error];
  }
  FBDeviceModel model = simDevice.deviceType.name;
  FBDeviceType *deviceType = FBiOSTargetConfiguration.nameToDevice[model];
  if (!deviceType) {
    return [[FBSimulatorError
      describeFormat:@"Could not obtain Device for for %@, perhaps it is unsupported by FBSimulatorControl", model]
      fail:error];
  }
  return [[FBSimulatorConfiguration.defaultConfiguration
    withOSNamed:osName]
    withDeviceModel:model];
}

+ (instancetype)inferSimulatorConfigurationFromDeviceSynthesizingMissing:(SimDevice *)simDevice
{
  FBSimulatorConfiguration *configuration = [self inferSimulatorConfigurationFromDevice:simDevice error:nil];
  if (configuration) {
    return configuration;
  }
  FBOSVersionName osName = simDevice.runtime.name;
  FBDeviceModel model = simDevice.deviceType.name;
  return [[FBSimulatorConfiguration.defaultConfiguration
    withOSNamed:osName]
    withDeviceModel:model];
}

- (BOOL)checkRuntimeRequirementsReturningError:(NSError **)error
{
  NSError *innerError = nil;
  SimRuntime *runtime = [self obtainRuntimeWithError:&innerError];
  if (!runtime) {
    return [[[FBSimulatorError
      describeFormat:@"Could not obtain available SimRuntime for configuration %@", self]
      causedBy:innerError]
      failBool:error];
  }
  SimDeviceType *deviceType = [self obtainDeviceTypeWithError:&innerError];
  if (!deviceType) {
    return [[[FBSimulatorError
      describeFormat:@"Could not obtain availableSimDeviceType for configuration %@", self]
      causedBy:innerError]
      failBool:error];
  }
  if (![runtime supportsDeviceType:deviceType]) {
    return [[FBSimulatorError
      describeFormat:@"Device Type %@ does not support Runtime %@", deviceType.name, runtime.name]
      failBool:error];
  }
  return YES;
}

+ (NSArray<FBSimulatorConfiguration *> *)allAvailableDefaultConfigurationsWithLogger:(nullable id<FBControlCoreLogger>)logger
{
  NSArray<NSString *> *absentOSVersions = nil;
  NSArray<NSString *> *absentDeviceTypes = nil;
  NSArray<FBSimulatorConfiguration *> *configurations = [self allAvailableDefaultConfigurationsWithAbsentOSVersionsOut:&absentOSVersions absentDeviceTypesOut:&absentDeviceTypes];
  for (NSString *absentOSVersion in absentOSVersions) {
    [logger.error logFormat:@"OS Version configuration for '%@' is missing", absentOSVersion];
  }
  for (NSString *absentDeviceType in absentDeviceTypes) {
    [logger.error logFormat:@"Device Type configuration for '%@' is missing", absentDeviceType];
  }
  return configurations;
}

+ (NSArray<FBOSVersion *> *)supportedOSVersions
{
  return [FBSimulatorConfiguration osVersionsForRuntimes:self.supportedRuntimes];
}

+ (NSArray<FBOSVersion *> *)supportedOSVersionsForDevice:(FBDeviceType *)device
{
  return [FBSimulatorConfiguration osVersionsForRuntimes:[self supportedRuntimesForDevice:device]];
}

+ (NSArray<FBSimulatorConfiguration *> *)allAvailableDefaultConfigurationsWithAbsentOSVersionsOut:(NSArray<NSString *> **)absentOSVersionsOut absentDeviceTypesOut:(NSArray<NSString *> **)absentDeviceTypesOut
{
  NSMutableArray<FBSimulatorConfiguration *> *configurations = [NSMutableArray array];
  NSMutableArray<NSString *> *absentOSVersions = [NSMutableArray array];
  NSMutableArray<NSString *> *absentDeviceTypes = [NSMutableArray array];
  NSArray<SimDeviceType *> *deviceTypes = self.supportedDeviceTypes;

  for (SimRuntime *runtime in self.supportedRuntimes) {
    if (!runtime.available) {
      continue;
    }
    FBOSVersionName osName = runtime.name;
    if (!FBiOSTargetConfiguration.nameToOSVersion[runtime.name]) {
      [absentOSVersions addObject:runtime.name];
      continue;
    }

    for (SimDeviceType *deviceType in deviceTypes) {
      if (![runtime supportsDeviceType:deviceType]) {
        continue;
      }
      FBDeviceModel model = deviceType.name;
      if (!FBiOSTargetConfiguration.nameToDevice[model]) {
        [absentDeviceTypes addObject:deviceType.name];
        continue;
      }

      FBSimulatorConfiguration *configuration = [[FBSimulatorConfiguration withDeviceModel:model] withOSNamed:osName];
      [configurations addObject:configuration];
    }
  }

  if (absentOSVersionsOut) {
    *absentOSVersionsOut = absentOSVersions;
  }
  if (absentDeviceTypesOut) {
    *absentDeviceTypesOut = absentDeviceTypes;
  }

  return [configurations copy];
}

#pragma mark Obtaining CoreSimulator Classes

- (SimRuntime *)obtainRuntimeWithError:(NSError **)error
{
  NSArray *supportedRuntimes = FBSimulatorConfiguration.supportedRuntimes;
  if (!supportedRuntimes) {
    return [[FBSimulatorError describe:@"Could not obtain supportedRuntimes, perhaps Framework loading failed"] fail:error];
  }
  NSArray *matchingRuntimes = [supportedRuntimes filteredArrayUsingPredicate:self.runtimePredicate];
  if (matchingRuntimes.count == 0) {
    return [[FBSimulatorError describeFormat:@"Could not obtain matching SimRuntime, no matches. Available Runtimes %@", supportedRuntimes] fail:error];
  }
  if (matchingRuntimes.count > 1) {
    return [[FBSimulatorError describeFormat:@"Matching Runtimes is ambiguous: %@", matchingRuntimes] fail:error];
  }
  return [matchingRuntimes firstObject];
}

- (SimDeviceType *)obtainDeviceTypeWithError:(NSError **)error
{
  NSArray *supportedDeviceTypes = FBSimulatorConfiguration.supportedDeviceTypes;
  if (!supportedDeviceTypes) {
    return [[FBSimulatorError describe:@"Could not obtain supportedDeviceTypes, perhaps Framework loading failed"] fail:error];
  }
  NSArray *matchingDeviceTypes = [supportedDeviceTypes filteredArrayUsingPredicate:[FBSimulatorConfiguration deviceTypePredicate:self.device]];
  if (matchingDeviceTypes.count == 0) {
    return [[FBSimulatorError describeFormat:@"Could not obtain matching DeviceTypes, no matches. Available Device Types %@", supportedDeviceTypes] fail:error];
  }
  if (matchingDeviceTypes.count > 1) {
    return [[FBSimulatorError describeFormat:@"Matching Device Types is ambiguous: %@", matchingDeviceTypes] fail:error];
  }
  return [matchingDeviceTypes firstObject];
}

#pragma mark Private

+ (NSArray<FBOSVersion *> *)osVersionsForRuntimes:(NSArray<SimRuntime *> *)runtimes
{
  NSMutableArray<FBOSVersion *> *array = [NSMutableArray array];
  for (SimRuntime *runtime in runtimes) {
    FBOSVersion *os = FBiOSTargetConfiguration.nameToOSVersion[runtime.name];
    if (!os) {
      os = [FBOSVersion genericWithName:runtime.name];
    }
    [array addObject:os];
  }
  return [array copy];
}

+ (NSArray<SimRuntime *> *)supportedRuntimes
{
  return FBSimulatorServiceContext.sharedServiceContext.supportedRuntimes;
}

+ (NSArray<SimDeviceType *> *)supportedDeviceTypes
{
  return FBSimulatorServiceContext.sharedServiceContext.supportedDeviceTypes;
}

+ (NSArray<SimRuntime *> *)supportedRuntimesForDevice:(FBDeviceType *)device
{
  return [[self.supportedRuntimes
    filteredArrayUsingPredicate:[FBSimulatorConfiguration runtimeProductFamilyPredicate:device]]
    sortedArrayUsingComparator:^ NSComparisonResult (SimRuntime *left, SimRuntime *right) {
      NSDecimalNumber *leftVersionNumber = [NSDecimalNumber decimalNumberWithString:left.versionString];
      NSDecimalNumber *rightVersionNumber = [NSDecimalNumber decimalNumberWithString:right.versionString];
      return [leftVersionNumber compare:rightVersionNumber];
    }];
}

- (NSPredicate *)runtimePredicate
{
  return [NSCompoundPredicate andPredicateWithSubpredicates:@[
    [FBSimulatorConfiguration runtimeProductFamilyPredicate:self.device],
    [FBSimulatorConfiguration runtimeNamePredicate:self.os],
    self.runtimeAvailabilityPredicate
  ]];
}

+ (NSPredicate *)runtimeProductFamilyPredicate:(FBDeviceType *)device
{
  return [NSPredicate predicateWithBlock:^ BOOL (SimRuntime *runtime, NSDictionary *_) {
    return [runtime.supportedProductFamilyIDs containsObject:@(device.family)];
  }];
}

+ (NSPredicate *)runtimeNamePredicate:(FBOSVersion *)OS
{
  return [NSPredicate predicateWithBlock:^ BOOL (SimRuntime *runtime, NSDictionary *_) {
    return [runtime.name isEqualToString:OS.name];
  }];
}

- (NSPredicate *)runtimeAvailabilityPredicate
{
  return [NSPredicate predicateWithBlock:^ BOOL (SimRuntime *runtime, NSDictionary *_) {
    return [runtime isAvailableWithError:nil];
  }];
}

+ (NSPredicate *)deviceTypePredicate:(FBDeviceType *)device
{
  return [NSPredicate predicateWithBlock:^ BOOL (SimDeviceType *deviceType, NSDictionary *_) {
    return [deviceType.name isEqualToString:device.model];
  }];
}

@end

File: FBSimulatorControl/Configuration/FBSimulatorConfiguration+CoreSimulator.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import CoreSimulator
@preconcurrency import FBControlCore
@preconcurrency import Foundation

extension FBSimulatorConfiguration {

  // MARK: - Matching Configuration against Available Versions

  public static func newestAvailableOS(forDevice device: FBDeviceType) throws -> FBOSVersion? {
    try FBSimulatorConfiguration.supportedOSVersions(forDevice: device).last
  }

  public func newestAvailableOS() throws -> FBSimulatorConfiguration {
    guard let os = try FBSimulatorConfiguration.newestAvailableOS(forDevice: device) else {
      throw FBSimulatorConfigurationError.noNewestAvailableOS(device: device.model.rawValue)
    }
    return withOSNamed(os.name)
  }

  public static func oldestAvailableOS(forDevice device: FBDeviceType) throws -> FBOSVersion? {
    try FBSimulatorConfiguration.supportedOSVersions(forDevice: device).first
  }

  public func oldestAvailableOS() throws -> FBSimulatorConfiguration {
    guard let os = try FBSimulatorConfiguration.oldestAvailableOS(forDevice: device) else {
      throw FBSimulatorConfigurationError.noOldestAvailableOS(device: device.model.rawValue)
    }
    return withOSNamed(os.name)
  }

  public static func inferSimulatorConfiguration(fromDevice simDevice: SimDevice) throws -> FBSimulatorConfiguration {
    let osName = FBOSVersionName(rawValue: simDevice.runtime.name ?? "unknown")
    guard FBiOSTargetConfiguration.nameToOSVersion[osName] != nil else {
      throw FBSimulatorConfigurationError.unsupportedOSVersion(name: osName.rawValue)
    }
    let model = FBDeviceModel(rawValue: simDevice.deviceType.name ?? "unknown")
    guard FBiOSTargetConfiguration.nameToDevice[model] != nil else {
      throw FBSimulatorConfigurationError.unsupportedDevice(name: model.rawValue)
    }
    return try FBSimulatorConfiguration.defaultConfiguration().withOSNamed(osName).withDeviceModel(model)
  }

  public static func inferSimulatorConfigurationFromDeviceSynthesizingMissing(_ simDevice: SimDevice) -> FBSimulatorConfiguration {
    if let configuration = try? inferSimulatorConfiguration(fromDevice: simDevice) {
      return configuration
    }
    // Synthesize directly rather than via the throwing `defaultConfiguration()`: this path must not
    // fail (it has ObjC callers in non-throwing FBSimulator init) and it overrides both OS and device
    // anyway, so the default's own values are irrelevant.
    let osName = FBOSVersionName(rawValue: simDevice.runtime.name ?? "unknown")
    let model = FBDeviceModel(rawValue: simDevice.deviceType.name ?? "unknown")
    let os = FBiOSTargetConfiguration.nameToOSVersion[osName] ?? FBOSVersion.generic(withName: osName.rawValue)
    let device = FBiOSTargetConfiguration.nameToDevice[model] ?? FBDeviceType.generic(withName: model.rawValue)
    return FBSimulatorConfiguration(device: device, os: os).withDeviceModel(model)
  }

  public func checkRuntimeRequirements() throws {
    let runtime: SimRuntime
    do {
      runtime = try obtainRuntime()
    } catch {
      throw FBSimulatorConfigurationError.runtimeUnavailable(configuration: "\(self)", reason: error.localizedDescription)
    }
    let deviceType: SimDeviceType
    do {
      deviceType = try obtainDeviceType()
    } catch {
      throw FBSimulatorConfigurationError.deviceTypeUnavailable(configuration: "\(self)", reason: error.localizedDescription)
    }
    if !runtime.supportsDeviceType(deviceType) {
      throw FBSimulatorConfigurationError.runtimeDeviceTypeMismatch(
        deviceType: deviceType.name ?? "unknown",
        runtime: runtime.name ?? "unknown")
    }
  }

  public static func supportedOSVersions() throws -> [FBOSVersion] {
    try osVersions(forRuntimes: supportedRuntimes())
  }

  public static func supportedOSVersions(forDevice device: FBDeviceType) throws -> [FBOSVersion] {
    try osVersions(forRuntimes: supportedRuntimes(forDevice: device))
  }

  public static func allAvailableDefaultConfigrations(withLogger logger: (any FBControlCoreLogger)?) throws -> [FBSimulatorConfiguration] {
    var absentOSVersions: NSArray?
    var absentDeviceTypes: NSArray?
    let configurations = try allAvailableDefaultConfigrations(withAbsentOSVersionsOut: &absentOSVersions, absentDeviceTypesOut: &absentDeviceTypes)
    if let absentOSVersions = absentOSVersions as? [String] {
      for osVersion in absentOSVersions {
        logger?.error().log("OS Version configuration for '\(osVersion)' is missing")
      }
    }
    if let absentDeviceTypes = absentDeviceTypes as? [String] {
      for deviceType in absentDeviceTypes {
        logger?.error().log("Device Type configuration for '\(deviceType)' is missing")
      }
    }
    return configurations
  }

  public static func allAvailableDefaultConfigrations(
    withAbsentOSVersionsOut absentOSVersionsOut: AutoreleasingUnsafeMutablePointer<NSArray?>?,
    absentDeviceTypesOut: AutoreleasingUnsafeMutablePointer<NSArray?>?
  ) throws -> [FBSimulatorConfiguration] {
    var configurations: [FBSimulatorConfiguration] = []
    var absentOSVersions: [String] = []
    var absentDeviceTypes: [String] = []
    let deviceTypes = try supportedDeviceTypes()

    for runtime in try supportedRuntimes() {
      if !runtime.available {
        continue
      }
      let runtimeName = runtime.name ?? "unknown"
      let osName = FBOSVersionName(rawValue: runtimeName)
      if FBiOSTargetConfiguration.nameToOSVersion[osName] == nil {
        absentOSVersions.append(runtimeName)
        continue
      }

      for deviceType in deviceTypes {
        if !runtime.supportsDeviceType(deviceType) {
          continue
        }
        let deviceTypeName = deviceType.name ?? "unknown"
        let model = FBDeviceModel(rawValue: deviceTypeName)
        if FBiOSTargetConfiguration.nameToDevice[model] == nil {
          absentDeviceTypes.append(deviceTypeName)
          continue
        }

        let configuration = try FBSimulatorConfiguration.defaultConfiguration().withDeviceModel(model).withOSNamed(osName)
        configurations.append(configuration)
      }
    }

    absentOSVersionsOut?.pointee = absentOSVersions as NSArray
    absentDeviceTypesOut?.pointee = absentDeviceTypes as NSArray
    return configurations
  }

  // MARK: - Obtaining CoreSimulator Classes

  public func obtainRuntime() throws -> SimRuntime {
    let runtimes = try FBSimulatorConfiguration.supportedRuntimes()
    let matchingRuntimes = runtimes.filter { runtime in
      runtime.available
        && runtime.name == os.name.rawValue
        && FBSimulatorConfiguration.runtime(runtime, supportsFamilyOf: device)
    }
    if matchingRuntimes.isEmpty {
      throw FBSimulatorConfigurationError.noMatchingRuntime(available: "\(runtimes)")
    }
    if matchingRuntimes.count > 1 {
      throw FBSimulatorConfigurationError.ambiguousRuntime(matches: "\(matchingRuntimes)")
    }
    return matchingRuntimes[0]
  }

  public func obtainDeviceType() throws -> SimDeviceType {
    let deviceTypes = try FBSimulatorConfiguration.supportedDeviceTypes()
    let matchingDeviceTypes = deviceTypes.filter { $0.name == device.model.rawValue }
    if matchingDeviceTypes.isEmpty {
      throw FBSimulatorConfigurationError.noMatchingDeviceType(available: "\(matchingDeviceTypes)")
    }
    if matchingDeviceTypes.count > 1 {
      throw FBSimulatorConfigurationError.ambiguousDeviceType(matches: "\(matchingDeviceTypes)")
    }
    return matchingDeviceTypes[0]
  }

  // MARK: - Private

  private static func osVersions(forRuntimes runtimes: [SimRuntime]) -> [FBOSVersion] {
    runtimes.map { runtime in
      guard let name = FBOSVersionName(rawValue: runtime.name) else {
        return FBOSVersion.generic(withName: "unknown")
      }
      return FBiOSTargetConfiguration.nameToOSVersion[name] ?? FBOSVersion.generic(withName: name.rawValue)
    }
  }

  private static func supportedRuntimes() throws -> [SimRuntime] {
    try FBSimulatorServiceContext.sharedServiceContext().supportedRuntimes()
  }

  private static func supportedDeviceTypes() throws -> [SimDeviceType] {
    try FBSimulatorServiceContext.sharedServiceContext().supportedDeviceTypes()
  }

  private static func supportedRuntimes(forDevice device: FBDeviceType) throws -> [SimRuntime] {
    try supportedRuntimes()
      .filter { runtime($0, supportsFamilyOf: device) }
      .sorted { left, right in
        let leftVersion = NSDecimalNumber(string: left.versionString)
        let rightVersion = NSDecimalNumber(string: right.versionString)
        return leftVersion.compare(rightVersion) == .orderedAscending
      }
  }

  private static func runtime(_ runtime: SimRuntime, supportsFamilyOf device: FBDeviceType) -> Bool {
    guard let familyIDs = runtime.supportedProductFamilyIDs as? [NSNumber] else {
      return false
    }
    return familyIDs.contains(NSNumber(value: device.family.rawValue))
  }
}

extension FBOSVersionName {
  /// CoreSimulator's un-annotated headers surface `SimRuntime.name` as optional; a runtime
  /// with no name has no version name, rather than a version name made from a placeholder.
  fileprivate init?(rawValue: String?) {
    guard let rawValue else {
      return nil
    }
    self.init(rawValue: rawValue)
  }
}
